### ISSUE TITLE
Sort within a sort does not set $a, $b

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7998]
Migrated from [rt.perl.org#36430](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=36430) (status was 'resolved')

Searchable as RT36430$

**Comment by p5pRT** [comment]
### From @btilly

#### Created by btilly@gmail.com



If a sort happens while a sort is happening\, $a and $b are not set\.
A co\-worker ran into this on Perl 5\.8\.7\.  The original example was a
lot more complex\, but here is a toy example that demonstrates the
problem\.

&nbsp;  \#\! /usr/bin/perl
&nbsp;  use strict;
&nbsp;  use Data&#8203;::Dumper;

&nbsp;  my @&#8203;list = sort \{ A&#8203;::min\(@&#8203;$a\) \<=> A&#8203;::min\(@&#8203;$b\) \}
&nbsp;    \[3\, 1\, 5\]\,
&nbsp;    \[2\, 4\]\,
&nbsp;    \[0\];

&nbsp;  print Dumper\(\\@&#8203;list\);

&nbsp;  package A;
&nbsp;  sub min \{
&nbsp;    my @&#8203;list = sort \{
&nbsp;      print STDERR "No \\$a\\n" unless defined\($a\);
&nbsp;      print STDERR "No \\$b\\n" unless defined\($b\);
&nbsp;      $a \<=> $b;
&nbsp;    \} @&#8203;\_;
&nbsp;    $list\[0\];
&nbsp;  \}

We worked around it by changing min to&#8203;:

&nbsp;  package A;
&nbsp;  sub min \{
&nbsp;    my $s = sub \($$\) \{
&nbsp;      my \($a\, $b\) = @&#8203;\_;
&nbsp;      print STDERR "No \\$a\\n" unless defined\($a\);
&nbsp;      print STDERR "No \\$b\\n" unless defined\($b\);
&nbsp;      $a \<=> $b;
&nbsp;    \};
&nbsp;    my @&#8203;list = sort $s @&#8203;\_;
&nbsp;    $list\[0\];
&nbsp;  \}

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.3:

Configured by Debian Project at Sun Feb 15 17:22:09 EST 2004.

Summary of my perl5 (revision 5.0 version 8 subversion 3) configuration:
  Platform:
    osname=linux, osvers=2.4.22-xfs+ti1211, archname=i386-linux-thread-multi
    uname='linux kosh 2.4.22-xfs+ti1211 #1 sat oct 25 10:11:37 est 2003 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i386-linux -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.3 -Dsitearch=/usr/local/lib/perl/5.8.3 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.3 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.3.3 20040125 (prerelease) (Debian)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=true, libperl=libperl.so.5.8.3
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.3:
    /etc/perl
    /usr/local/lib/perl/5.8.3
    /usr/local/share/perl/5.8.3
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    /usr/local/lib/perl/5.8.2
    /usr/local/share/perl/5.8.2
    /usr/local/lib/perl/5.8.1
    /usr/local/share/perl/5.8.1
    /usr/local/lib/perl/5.8.0
    /usr/local/share/perl/5.8.0
    .


Environment for perl v5.8.3:
    HOME=/home/btilly
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/usr/local/oracle/lib
    LOGDIR (unset)
    PATH=/home/btilly/cvs/bin:/home/btilly/cvs/lib/oracle/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games:/usr/share/games:~/cvs/bin:/usr/local/oracle/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From guest@guest.guest.xxxxxxxx

> \[btilly@&#8203;gmail\.com \- Wed Jun 29 17&#8203;:56&#8203;:08 2005\]&#8203;:
> If a sort happens while a sort is happening\, $a and $b are not set\.
> A co\-worker ran into this on Perl 5\.8\.7\.  The original example was a
> lot more complex\, but here is a toy example that demonstrates the
> problem\.
> 
>   \#\! /usr/bin/perl
>   use strict;
>   use Data&#8203;::Dumper;
> 
>   my @&#8203;list = sort \{ A&#8203;::min\(@&#8203;$a\) \<=> A&#8203;::min\(@&#8203;$b\) \}
>     \[3\, 1\, 5\]\,
>     \[2\, 4\]\,
>     \[0\];
> 
>   print Dumper\(\\@&#8203;list\);
> 
>   package A;
>   sub min \{
>     my @&#8203;list = sort \{
>       print STDERR "No \\$a\\n" unless defined\($a\);
>       print STDERR "No \\$b\\n" unless defined\($b\);
>       $a \<=> $b;
>     \} @&#8203;\_;
>     $list\[0\];
>   \}

I think this is a duplicate of bug 7063\, and the problem is actually
caused by the fact that the sort\(\) is in a different package from the
sort sub\.

I know how to fix it\, but I'm slightly concerned that people might be
relying on the existing \(mis\)behaviour\, which is that $a and $b are set
in the package that the sort is called from instead of the package where
the comparison sub lives\.

\(The existing behaviour does seem to be accidental\. The code looks as
though it intends to use the package of the comparison routine\, but gets
it a bit wrong\.\)

Robin

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @btilly

On 10/25/05\, Guest via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > \[btilly@&#8203;gmail\.com \- Wed Jun 29 17&#8203;:56&#8203;:08 2005\]&#8203;:
> > If a sort happens while a sort is happening\, $a and $b are not set\.
> > A co\-worker ran into this on Perl 5\.8\.7\.  The original example was a
> > lot more complex\, but here is a toy example that demonstrates the
> > problem\.
> >
> >   \#\! /usr/bin/perl
> >   use strict;
> >   use Data&#8203;::Dumper;
> >
> >   my @&#8203;list = sort \{ A&#8203;::min\(@&#8203;$a\) \<=> A&#8203;::min\(@&#8203;$b\) \}
> >     \[3\, 1\, 5\]\,
> >     \[2\, 4\]\,
> >     \[0\];
> >
> >   print Dumper\(\\@&#8203;list\);
> >
> >   package A;
> >   sub min \{
> >     my @&#8203;list = sort \{
> >       print STDERR "No \\$a\\n" unless defined\($a\);
> >       print STDERR "No \\$b\\n" unless defined\($b\);
> >       $a \<=> $b;
> >     \} @&#8203;\_;
> >     $list\[0\];
> >   \}
>
> I think this is a duplicate of bug 7063\, and the problem is actually
> caused by the fact that the sort\(\) is in a different package from the
> sort sub\.

I believe that you do not fully understand the problem\.  There are 2
sorts\, and for each sort\, the sort and the comparison function are in
the same package\.  There is no mismatch there\.  If you call the min
function\, it normally works just fine\.

The problem is that the inner sort is setting $a and $b in the right
package for the OUTER sort\.  So min fails because the sort called in
package A with a comparison function in package A is setting $main&#8203;::a
and $main&#8203;::b\.  That's the bug\.

> I know how to fix it\, but I'm slightly concerned that people might be
> relying on the existing \(mis\)behaviour\, which is that $a and $b are set
> in the package that the sort is called from instead of the package where
> the comparison sub lives\.

People do rely on the existing behaviour\, and it is documented that if
you wish to use a sort function from a different package than the sort
is called in\, that you should use a \($$\) prototype\.  I agree that it
would be nicer if $a and $b were set in the comparison subroutine's
package\, but backwards compatibility wins\.  \(Setting $a and $b in
\_both\_ packages would be a nice piece of DWIMery\.\.\.\)

> \(The existing behaviour does seem to be accidental\. The code looks as
> though it intends to use the package of the comparison routine\, but gets
> it a bit wrong\.\)

The existing behaviour may have been a bug\, but it has seniority
\(people depend on it\) and so is now a feature\.  What is NOT a feature
\(it is probably as old\, but nobody depends on it\) is that if sort is
called while sort is already being called\, the package in which $a and
$b are set by the current sort is determined by the package of the
initial sort\.

Cheers\,
Ben

**Comment by p5pRT** [comment]
### From guest@guest.guest.xxxxxxxx

\[btilly@&#8203;gmail\.com \- Wed Oct 26 01&#8203;:43&#8203;:11 2005\]&#8203;:
> I believe that you do not fully understand the problem\.  There are 2
> sorts\, and for each sort\, the sort and the comparison function are in
> the same package\.  There is no mismatch there\.  If you call the min
> function\, it normally works just fine\.

Right you are\. Sorry for the confusion\!

Robin

**Comment by p5pRT** [comment]
### From @smpeters

> \[btilly@&#8203;gmail\.com \- Wed Jun 29 17&#8203;:56&#8203;:08 2005\]&#8203;:
> 
> 
> This is a bug report for perl from btilly@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.34 running under perl v5\.8\.3\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> If a sort happens while a sort is happening\, $a and $b are not set\.
> A co\-worker ran into this on Perl 5\.8\.7\.  The original example was a
> lot more complex\, but here is a toy example that demonstrates the
> problem\.
> 
>   \#\! /usr/bin/perl
>   use strict;
>   use Data&#8203;::Dumper;
> 
>   my @&#8203;list = sort \{ A&#8203;::min\(@&#8203;$a\) \<=> A&#8203;::min\(@&#8203;$b\) \}
>     \[3\, 1\, 5\]\,
>     \[2\, 4\]\,
>     \[0\];
> 
>   print Dumper\(\\@&#8203;list\);
> 
>   package A;
>   sub min \{
>     my @&#8203;list = sort \{
>       print STDERR "No \\$a\\n" unless defined\($a\);
>       print STDERR "No \\$b\\n" unless defined\($b\);
>       $a \<=> $b;
>     \} @&#8203;\_;
>     $list\[0\];
>   \}
> 
> We worked around it by changing min to&#8203;:
> 
>   package A;
>   sub min \{
>     my $s = sub \($$\) \{
>       my \($a\, $b\) = @&#8203;\_;
>       print STDERR "No \\$a\\n" unless defined\($a\);
>       print STDERR "No \\$b\\n" unless defined\($b\);
>       $a \<=> $b;
>     \};
>     my @&#8203;list = sort $s @&#8203;\_;
>     $list\[0\];
>   \}
> 

With the following patch\, the warnings for $a and $b not being defined 
go away\, but I'm not sure if its correct or not\.  I'd appreciate 
someone more knowledgible with than myself on the Perl internals taking 
a look at it\.  Essentially\, I noticed that Perl\_pp\_sort\(\) contains a 
SAVETMPS\, but not a matching FREETMPS\.  Adding the FREETMPS\, makes the 
internal sort happy and produces identical results workaround\.

<details><summary>Inline Patch</summary>

```diff
--- pp_sort.c.old       2005-12-14 10:51:02.000000000 -0600
+++ pp_sort.c   2005-12-14 09:49:16.000000000 -0600
@@ -1686,6 +1686,7 @@
            PL_stack_sp = newsp;
            POPSTACK;
            CATCH_SET(oldcatch);
+           FREETMPS;
        }
        else {
            MEXTEND(SP, 20);    /* Can't afford stack realloc on 
signal. */
```
</details>

**Comment by p5pRT** [comment]
### From @smpeters

> \[stmpeters \- Wed Dec 14 08&#8203;:51&#8203;:40 2005\]&#8203;:
> 
> > \[btilly@&#8203;gmail\.com \- Wed Jun 29 17&#8203;:56&#8203;:08 2005\]&#8203;:
> > 
> > 
> > This is a bug report for perl from btilly@&#8203;gmail\.com\,
> > generated with the help of perlbug 1\.34 running under perl v5\.8\.3\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> > 
> > If a sort happens while a sort is happening\, $a and $b are not set\.
> > A co\-worker ran into this on Perl 5\.8\.7\.  The original example was a
> > lot more complex\, but here is a toy example that demonstrates the
> > problem\.
> > 
> >   \#\! /usr/bin/perl
> >   use strict;
> >   use Data&#8203;::Dumper;
> > 
> >   my @&#8203;list = sort \{ A&#8203;::min\(@&#8203;$a\) \<=> A&#8203;::min\(@&#8203;$b\) \}
> >     \[3\, 1\, 5\]\,
> >     \[2\, 4\]\,
> >     \[0\];
> > 
> >   print Dumper\(\\@&#8203;list\);
> > 
> >   package A;
> >   sub min \{
> >     my @&#8203;list = sort \{
> >       print STDERR "No \\$a\\n" unless defined\($a\);
> >       print STDERR "No \\$b\\n" unless defined\($b\);
> >       $a \<=> $b;
> >     \} @&#8203;\_;
> >     $list\[0\];
> >   \}
> > 
> > We worked around it by changing min to&#8203;:
> > 
> >   package A;
> >   sub min \{
> >     my $s = sub \($$\) \{
> >       my \($a\, $b\) = @&#8203;\_;
> >       print STDERR "No \\$a\\n" unless defined\($a\);
> >       print STDERR "No \\$b\\n" unless defined\($b\);
> >       $a \<=> $b;
> >     \};
> >     my @&#8203;list = sort $s @&#8203;\_;
> >     $list\[0\];
> >   \}
> > 
> 
> With the following patch\, the warnings for $a and $b not being defined 
> go away\, but I'm not sure if its correct or not\.  I'd appreciate 
> someone more knowledgible with than myself on the Perl internals taking 
> a look at it\.  Essentially\, I noticed that Perl\_pp\_sort\(\) contains a 
> SAVETMPS\, but not a matching FREETMPS\.  Adding the FREETMPS\, makes the 
> internal sort happy and produces identical results workaround\.
> 
> \-\-\- pp\_sort\.c\.old       2005\-12\-14 10&#8203;:51&#8203;:02\.000000000 \-0600
> \+\+\+ pp\_sort\.c   2005\-12\-14 09&#8203;:49&#8203;:16\.000000000 \-0600
> @&#8203;@&#8203; \-1686\,6 \+1686\,7 @&#8203;@&#8203;
>             PL\_stack\_sp = newsp;
>             POPSTACK;
>             CATCH\_SET\(oldcatch\);
> \+           FREETMPS;
>         \}
>         else \{
>             MEXTEND\(SP\, 20\);    /\* Can't afford stack realloc on 
> signal\. \*/
> 

The current tests all work with this fix\, including threaded and DEBUGGING\.  So\, I've applied 
this as change \#26361 \(and \#26360\, oops\)\.

**Comment by p5pRT** [comment]
### From @smpeters

> \[stmpeters \- Wed Dec 14 08&#8203;:51&#8203;:40 2005\]&#8203;:
> 
> > \[btilly@&#8203;gmail\.com \- Wed Jun 29 17&#8203;:56&#8203;:08 2005\]&#8203;:
> > 
> > 
> > This is a bug report for perl from btilly@&#8203;gmail\.com\,
> > generated with the help of perlbug 1\.34 running under perl v5\.8\.3\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> > 
> > If a sort happens while a sort is happening\, $a and $b are not set\.
> > A co\-worker ran into this on Perl 5\.8\.7\.  The original example was a
> > lot more complex\, but here is a toy example that demonstrates the
> > problem\.
> > 
> >   \#\! /usr/bin/perl
> >   use strict;
> >   use Data&#8203;::Dumper;
> > 
> >   my @&#8203;list = sort \{ A&#8203;::min\(@&#8203;$a\) \<=> A&#8203;::min\(@&#8203;$b\) \}
> >     \[3\, 1\, 5\]\,
> >     \[2\, 4\]\,
> >     \[0\];
> > 
> >   print Dumper\(\\@&#8203;list\);
> > 
> >   package A;
> >   sub min \{
> >     my @&#8203;list = sort \{
> >       print STDERR "No \\$a\\n" unless defined\($a\);
> >       print STDERR "No \\$b\\n" unless defined\($b\);
> >       $a \<=> $b;
> >     \} @&#8203;\_;
> >     $list\[0\];
> >   \}
> > 
> > We worked around it by changing min to&#8203;:
> > 
> >   package A;
> >   sub min \{
> >     my $s = sub \($$\) \{
> >       my \($a\, $b\) = @&#8203;\_;
> >       print STDERR "No \\$a\\n" unless defined\($a\);
> >       print STDERR "No \\$b\\n" unless defined\($b\);
> >       $a \<=> $b;
> >     \};
> >     my @&#8203;list = sort $s @&#8203;\_;
> >     $list\[0\];
> >   \}
> > 
> 
> With the following patch\, the warnings for $a and $b not being defined 
> go away\, but I'm not sure if its correct or not\.  I'd appreciate 
> someone more knowledgible with than myself on the Perl internals taking 
> a look at it\.  Essentially\, I noticed that Perl\_pp\_sort\(\) contains a 
> SAVETMPS\, but not a matching FREETMPS\.  Adding the FREETMPS\, makes the 
> internal sort happy and produces identical results workaround\.
> 
> \-\-\- pp\_sort\.c\.old       2005\-12\-14 10&#8203;:51&#8203;:02\.000000000 \-0600
> \+\+\+ pp\_sort\.c   2005\-12\-14 09&#8203;:49&#8203;:16\.000000000 \-0600
> @&#8203;@&#8203; \-1686\,6 \+1686\,7 @&#8203;@&#8203;
>             PL\_stack\_sp = newsp;
>             POPSTACK;
>             CATCH\_SET\(oldcatch\);
> \+           FREETMPS;
>         \}
>         else \{
>             MEXTEND\(SP\, 20\);    /\* Can't afford stack realloc on 
> signal\. \*/
> 

Wait\, my change did nothing\.  This appears to have been fixed already by change \#25953\.  
Time to revert my change\(s\)\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
